packages/bot/src/scratch/hooks/toolbox.js

```diff
@@ -80,122 +80,135 @@ Blockly.Toolbox.prototype.populate_ = function (newTree) {
+
+                if (definition instanceof Array) {
+                    let has_dropdown_and_in_search = false;
+                    // eslint-disable-next-line consistent-return
                     definition.forEach(def => {
-                        if (def instanceof Object) {
-                            keywords += !def.type.includes('image') ? ` ${JSON.stringify(def)}` : '';
-                        } else {
-                            keywords += ` ${def}`;
+                        if (def.type === 'field_dropdown' &&
+                            JSON.stringify(def).toUpperCase().includes(search_term)) {
+                            has_dropdown_and_in_search = true;
                         }
                     });
+
+                    return has_dropdown_and_in_search;
                 }
-            });
 
-            const block_category = (block_definition && block_definition.category) ||
-                                    (block_meta && block_meta.category);
-            const category =
-                this.categoryMenu_.categories_
-                    .find(menuCategory => menuCategory.id_ === block_category);
-            let contents = category && category.getContents();
+                return false;
+            });
+            if (matched_definition && matched_definition.length) {
+                flyout_content.push(block_content);
 
-            if (typeof contents === 'string') {
-                const fnToApply = this.workspace_.getToolboxCategoryCallback(contents);
-                contents = fnToApply(this.workspace_);
             }
+        });
 
-            if (contents) {
-                search_term.forEach(term => {
-                    if (keywords.toLowerCase().includes(term)) {
-                        const blockContents = contents
-                            .filter(content => content.attributes[0].nodeValue === blockKey);
+        // block_variable_name matched
+        const matched_variables = all_variables
+            .filter(variable => variable.name.toUpperCase().includes(search_term));
+        const variables_blocks = Blockly.DataCategory.search(matched_variables);
+        // eslint-disable-next-line consistent-return
+        const uniqueVarBlocks = variables_blocks.filter(variable_block => {
+            return flyout_content.indexOf(variable_block) === -1;
+        });
+        if (uniqueVarBlocks && uniqueVarBlocks.length) {
+            flyout_content.push(...uniqueVarBlocks);
+        }
 
-                        if (blockContents.length && flyout_content.blocks.indexOf(blockContents[0]) === -1) {
-                            flyout_content.blocks.push(blockContents[0]);
-                        }
+        // block_procedure_name matched
+        const searched_procedures = { '0': [], '1': [] };
+        const procedures_callnoreturn = all_procedures[0];
+        const procedures_callreturn = all_procedures[1];
+        Object.keys(procedures_callnoreturn).forEach(key => {
```

Comment:
The logic for looping through `procedures_callnoreturn` and `procedures_callreturn` is the same, you can create a single loop for this.